### PR TITLE
Respawn delay

### PR DESCRIPTION
**Pull request by contradict**
Sometimes, especially for hardware drivers, it helps to pause before restarting a failed node.

This branch adds a configurable delay to the roslaunch respawn feature. If the delay is left unspecified, it defaults to 0, giving the same behavior as before. The value is configured in the node tag:
<node
     name="myname"
     pkg="flaky_hardware"
     type="unreliable_node"
    respawn="True"
    respawn_delay="5.0"
/>
respawn_delay has units seconds.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder

**Comment by dirk-thomas** [comment]
Please update the PR so that added/changed lines comply with PEP 8 and avoid any non-necessary whitespace changes in order to keep the diff minimal.

**Comment by contradict** [comment]
Can you be more specific? I took a crack at this and found that most of what I could do was breaking long lines. Most of the existing code uses very long lines, so I'm not sure this is a good change. I found one line where I unnecessarily changed the indentation and a couple where I used space inside parens in a non-pep 8 way. Is there some aspect of the formatting you would particularly like me to fix?

File: tools/roslaunch/src/roslaunch/pmon.py [hunk 1]
```diff
@@ -538,15 +558,15 @@ def _run(self):
             for p in procs:
                 try:
                     if not p.is_alive():
-                        logger.debug("Process[%s] has died, respawn=%s, required=%s, exit_code=%s",p.name, p.respawn, p.required, p.exit_code)
+                        logger.debug("Process[%s] has died, respawn=%s, required=%s, exit_code=%s",
+                                p.name,
+                                "True(%f)"%p.respawn_delay if p.respawn else p.respawn,
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/pmon.py]: Please add spaces around `%`.

File: tools/roslaunch/src/roslaunch/pmon.py [hunk 2]
```diff
@@ -582,18 +604,23 @@ def _run(self):
...
+            _respawn=[]
             for r in respawn: 
                 try:
                     if self.is_shutdown:
                         break
-                    printlog("[%s] restarting process"%r.name)
-                    # stop process, don't accumulate errors
-                    r.stop([])
-                    r.start()
+                    if r.should_respawn()<=0.0:
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/pmon.py]: Please add spaces around `<=`.

File: tools/roslaunch/src/roslaunch/pmon.py [hunk 3]
```diff
@@ -582,18 +604,23 @@ def _run(self):
...
+                        printlog("[%s] restarting process"%r.name)
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/pmon.py]: Spaces around `%`.

File: tools/roslaunch/test/unit/test_roslaunch_pmon.py [hunk 4]
```diff
@@ -427,6 +462,12 @@ def f():
         
         self.failIf(marker.marked, "pmon had to be externally killed")        
 
+        self.failIf(p3.spawn_count < 2, "process did not respawn")
+
+        self.failIf(p4.respawn_interval < p4.respawn_delay,
+                "Respawn delay not respected: %s %s"%(p4.respawn_interval,
```
Comment (dirk-thomas) [inline tools/roslaunch/test/unit/test_roslaunch_pmon.py]: Spaces around `%`.

File: tools/roslaunch/src/roslaunch/core.py [hunk 5]
```diff
@@ -514,6 +518,7 @@ def xmlattrs(self):
             ('output', self.output),
             ('cwd', cwd_str), 
             ('respawn', self.respawn), #not valid on <test>
+            ('respawn_delay', self.respawn_delay), #not valid on <test>
```
Comment (dirk-thomas) [inline tools/roslaunch/src/roslaunch/core.py]: A trailing comment should have two spaces before the `#` and one after.

**Comment by dirk-thomas** [comment]
Please rebase the PR to the tip of the branch.

**Comment by contradict** [comment]
I have so far been unable to duplicate this build failure on my system. I tried running the tests using the jenkins_tools package, setting up my own containerized environment, and running the test in my development environment. In all cases, the test which fails here runs successfully. The raw log of the test run above indeed looks different than a successful run, it emits the same error message as a successful run but much more frequently. What can I do to more accurately replicate the test environment?

**Comment by dirk-thomas** [comment]
You can only run the tests locally (`make run_tests`). There is currently no good way to resemble the exact environment from the build farm.

I will check if I see anything obvious but might not have enough time to get to a useful result.

**Comment by dirk-thomas** [comment]
@ros-pull-request-builder

**Comment by wjwwood** [comment]
+1

**Comment by dirk-thomas** [comment]
Squashed and merged in c171283e548cfdbe5bdafeb5559462beff529a89

**Comment by dirk-thomas** [comment]
@contradict Can you please update the documentation in the wiki to mention the new attribute introduced in _Indigo_ (and post the link to the wiki changes here)?

**Comment by contradict** [comment]
Update added here, I hope this is the correct place: http://wiki.ros.org/roslaunch/XML/node

**Comment by dirk-thomas** [comment]
Thanks, I have updated the page to mention that the attribute has been introduced in Indigo: http://wiki.ros.org/action/diff/roslaunch/XML/node?action=diff&rev1=20&rev2=21

**Pull request closed without merging**